Single-cell CRISPR screens (e.g., perturb-seq) combine CRISPR genome engineering and single-cell sequencing to survey the effects of genetic perturbations on individual cells. Single-cell CRISPR screens pose enormous potential for unraveling mechanisms underlying disease and accelerating drug discovery. However, the analysis of these screens presents considerable statistical and computational challenges. *Applied Single-Cell CRISPR Screen Data Analysis* is a step-by-step guide to carrying out statistically rigorous and computationally efficient single-cell CRISPR screen data analysis using the `sceptre` R package. `sceptre` was developed by [Timothy Barry](https://timothy-barry.github.io/), Joseph Deutsch, and [Eugene Katsevich](https://ekatsevi.github.io/).

`sceptre` is compatible with a broad range of single-cell CRISPR screen experimental designs. In particular, datasets can vary along the following axes: multiplicity of infection (low or high), genomic element targeted (gene or noncoding regulatory element), CRISPR modality (CRISPRko, CRISPRi, CRISPRa, or CRISPR prime/base editing), and molecular readout (gene expression, protein expression, or chromatin accessibility). This book uses two real datasets to illustrate use of `sceptre`: a high-MOI CRISPRi screen of noncoding regulatory elements and a low-MOI CRISPRko screen of gene transcription start sites. On both datasets the molecular phenotype measured is the transcriptome (i.e., the gene expressions).

# Navigating this Book {.unnumbered}

Users should start by reading the *The whole game* (link below), which provides a high-level overview of `sceptre`. Users who wish to learn more about individual steps of the `sceptre` pipeline should read the subsequent chapters (i.e., *Import data* to *Run power check and discovery analysis*) in order.
