# Welcome {.unnumbered}

Single-cell CRISPR screens (e.g., perturb-seq) combine CRISPR genome engineering and single-cell sequencing to survey the effects of genetic perturbations on individual cells. Single-cell CRISPR screens pose enormous potential for unraveling mechanisms underlying disease and accelerating drug discovery. However, the analysis of these screens presents considerable statistical and computational challenges. *Hands-On Single-Cell CRISPR Screen Analysis* is a step-by-step guide to carrying out statistically rigorous and massively scalable single-cell CRISPR screen data analysis using the [sceptre](https://katsevich-lab.github.io/sceptre/) R package.

`sceptre` is compatible with a broad range of single-cell CRISPR screen experimental designs. In particular, datasets can vary along the following axes: multiplicity of infection (low or high), genomic element targeted (gene or noncoding regulatory element), CRISPR modality (CRISPRko, CRISPRi, CRISPRa, or CRISPR prime or base editing), and molecular readout (gene expression, protein expression, or chromatin accessibility). Additionally, `sceptre` enables the analysis of massive-scale data that may be too large to fit into memory or that require a cluster or cloud to process.

# Navigating this book {.unnumbered}

This book contains an introductory chapter and three parts.

-   [The whole game](sceptre.qmd) provides a high-level overview of `sceptre`. First-time users can get started by reading this chapter only.

-   Part I ([Chapter 1](import-data.qmd) --- [Chapter 6](run-power-check-and-discovery-analysis.qmd)) describes in greater detail how to carry out a `sceptre` analysis on a laptop or desktop.

-   Part II ([Chapter 7](at-scale.qmd) --- [Chapter 9](pipeline-args.qmd)) describes how leverage `sceptre` in conjunction with the companion R package [`ondisc`](https://timothy-barry.github.io/ondisc/) to analyze data at-scale on a cluster or cloud.

-   Part III ([Chapter 10](methods.qmd) --- ongoing) explicates the methods and algorithms that underlie `sceptre`.

Finally, the [Appendix](appendix.qmd) contains several sections, including a glossary and FAQs.

# Acknowledgments {.unnumbered}

We gratefully acknowledge Kathryn Roeder, Xihong Lin,  Xuran Wang, John Morris, Kaishu Mason, Ziang Niu, Yixuan Qiu, and Songcheng Dai for contributing to the research and development underlying `sceptre` and `ondisc`. We additionally thank the many `sceptre` users who have provided and continue to provide valuable feedback on the package.
