Single-cell CRISPR screens (e.g., perturb-seq) combine CRISPR genome engineering and single-cell sequencing to survey the effects of genetic perturbations on individual cells. Single-cell CRISPR screens pose enormous potential for unraveling mechanisms underlying disease and accelerating drug discovery. However, the analysis of these screens poses considerable statistical and computational challenges. *Applied Single-Cell CRISPR Screen Data Analysis* is a step-by-step guide to statistically rigorous, computationally efficient, and user-friendly single-cell CRISPR screen data analysis using the `sceptre` R package. `sceptre` was developed by Timothy Barry, Joseph Deutsch, and Eugene Katsevich.

# Using this Book {.unnumbered}

All users should read the opening chapter (i.e., *The whole game*), and advanced users should read the subsequent chapters (i.e., *Import data* to *Run power check and discovery analysis*) in order.

# Acknowledgements {.unnumbered}

We gratefully acknowledge Kathryn Roeder, John Morris, Kaishu Mason, Xuran Wang, and Ziang Niu for contributing to the research and development underlying `sceptre`. We additionally thank the many `sceptre` users who have provided valuable feedback on early versions of the package.
