<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Timothy Barry, Eugene Katsevich">
<title>Hands-On Single-Cell CRISPR Screen Analysis - 7&nbsp; Overview of at-scale sceptre</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./pipeline-details.html" rel="next">
<link href="./run-power-check-and-discovery-analysis.html" rel="prev">
<link href="./book_cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="custom.css">
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./at-scale.html">Part II. At-scale sceptre</a></li><li class="breadcrumb-item"><a href="./at-scale.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Overview of at-scale `sceptre`</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./hex.jpg" alt="" class="sidebar-logo py-0 d-lg-inline d-none"></a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Hands-On Single-Cell CRISPR Screen Analysis</a> 
        <div class="sidebar-tools-main">
    <a href="https://katsevich-lab.github.io/sceptre/" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-git"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sceptre.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The whole game</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Part I. Introduction to sceptre</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./import-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Import data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./set-analysis-parameters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Set analysis parameters</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assign-grnas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Assign gRNAs</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./run-qc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Run quality control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./run-calibration-check.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Run calibration check</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./run-power-check-and-discovery-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Run power check and discovery analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Part II. At-scale sceptre</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./at-scale.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Overview of at-scale <code>sceptre</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipeline-details.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Details of the pipeline</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Part III. Methods and algorithms</span></span>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./glossary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Glossary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Frequently asked questions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">References</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#sec-import_data_into_odm_backed_sceptre_object" id="toc-sec-import_data_into_odm_backed_sceptre_object" class="nav-link active" data-scroll-target="#sec-import_data_into_odm_backed_sceptre_object"><span class="header-section-number">7.1</span> Import data into an <code>ondisc</code>-backed <code>sceptre</code> object (local machine)</a></li>
  <li><a href="#explore-the-data-local-machine" id="toc-explore-the-data-local-machine" class="nav-link" data-scroll-target="#explore-the-data-local-machine"><span class="header-section-number">7.2</span> Explore the data (local machine)</a></li>
  <li><a href="#set-the-analysis-parameters-local-machine" id="toc-set-the-analysis-parameters-local-machine" class="nav-link" data-scroll-target="#set-the-analysis-parameters-local-machine"><span class="header-section-number">7.3</span> Set the analysis parameters (local machine)</a></li>
  <li><a href="#write-the-ondisc-backed-sceptre_object-to-disk-local-machine" id="toc-write-the-ondisc-backed-sceptre_object-to-disk-local-machine" class="nav-link" data-scroll-target="#write-the-ondisc-backed-sceptre_object-to-disk-local-machine"><span class="header-section-number">7.4</span> Write the <code>ondisc</code>-backed <code>sceptre_object</code> to disk (local machine)</a></li>
  <li><a href="#call-the-trial-nextflow-pipeline-local-machine" id="toc-call-the-trial-nextflow-pipeline-local-machine" class="nav-link" data-scroll-target="#call-the-trial-nextflow-pipeline-local-machine"><span class="header-section-number">7.5</span> Call the trial Nextflow pipeline (local machine)</a></li>
  <li><a href="#sec-push_data_to_cluster" id="toc-sec-push_data_to_cluster" class="nav-link" data-scroll-target="#sec-push_data_to_cluster"><span class="header-section-number">7.6</span> Push the data to the cluster</a></li>
  <li><a href="#call-the-trial-pipeline-cluster" id="toc-call-the-trial-pipeline-cluster" class="nav-link" data-scroll-target="#call-the-trial-pipeline-cluster"><span class="header-section-number">7.7</span> Call the trial pipeline (cluster)</a></li>
  <li><a href="#call-the-full-pipeline-cluster" id="toc-call-the-full-pipeline-cluster" class="nav-link" data-scroll-target="#call-the-full-pipeline-cluster"><span class="header-section-number">7.8</span> Call the full pipeline (cluster)</a></li>
  <li><a href="#stepping-through-the-pipeline-incrementally" id="toc-stepping-through-the-pipeline-incrementally" class="nav-link" data-scroll-target="#stepping-through-the-pipeline-incrementally"><span class="header-section-number">7.9</span> Stepping through the pipeline incrementally</a></li>
  <li><a href="#sec-install_nextflow" id="toc-sec-install_nextflow" class="nav-link" data-scroll-target="#sec-install_nextflow"><span class="header-section-number">7.10</span> Addendum: installing Nextflow</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="at-scale" class="quarto-section-identifier"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Overview of at-scale <code>sceptre</code></span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Part II of this book describes how to apply <code>sceptre</code> to analyze large-scale single-cell CRISPR screen data. Part II is slightly more advanced than Part I. Ideally, readers will have intermediate bioinformatics experience, including familiarity with basic linux commands and facility with submitting and monitoring jobs on a cluster. Knowledge of Nextflow (or some similar workflow management tool) is helpful not required.</p>
</div>
</div>
<p>This chapter describes how to apply <code>sceptre</code> to analyze large-scale single-cell CRISPR screen data. We begin by installing <code>ondisc</code>, a companion R package to <code>sceptre</code>. <code>ondisc</code> provides facilities for out-of-core and distributed computing on large-scale single-cell CRISPR screen data.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-1_bfeeb769e257a427e78e3c310b6adfeb">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"timothy-barry/ondisc"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Users can leverage <code>sceptre</code> and <code>ondisc</code> to analyze their data in one of two ways. First, users can analyze their data <em>out-of-core</em> on a laptop or desktop within the R console. This solution is most appropriate when the data are too big to fit into memory but the number of perturbation-gene pairs to be analyzed is not too large. Second, users can distribute their analysis across tens or even hundreds of processors on a computing cluster via the <code>sceptre</code> Nextflow pipeline. This solution is most appropriate when the data are big <em>and</em> a large number of pairs is to be analyzed. The addendum to this chapter (<a href="#sec-install_nextflow"><span>Section&nbsp;7.10</span></a>) explains how to download and install Nextflow.</p>
<p>The schematic below illustrates the two at-scale interfaces to <code>sceptre</code>. The first step is the same regardless of whether one plans to analyze the data within the R console or via the Nextflow pipeline: one imports the data into an <code>ondisc</code>-backed <code>sceptre_object</code>, which is a kind of <code>sceptre_object</code> in which the response and gRNA count matrices are stored on-disk rather than in memory. If one plans to analyze the data entirely within the R console, then one proceeds in the manner as described in <a href="sceptre.html#sec-whole_game_import_data"><span>Section&nbsp;1</span></a> — <a href="sceptre.html#sec-sceptre_write_outputs_to_directory"><span>Section&nbsp;8</span></a> of <em>The whole game</em>, calling the standard pipeline functions (i.e., <code><a href="https://rdrr.io/pkg/sceptre/man/assign_grnas.html">assign_grnas()</a></code>, <code><a href="https://rdrr.io/pkg/sceptre/man/run_qc.html">run_qc()</a></code>, etc.) on the <code>ondisc</code>-backed <code>sceptre_object</code> one-by-one. If, one the other hand, one plans to analyze the data via the Nextflow pipeline, then one proceeds through a slightly more involved (though still straightforward) sequence of steps. This chapter focuses on the Nextflow interface.</p>
<div class="cell" data-layout-align="center" data-hash="at-scale_cache/html/unnamed-chunk-2_9454315b4f3095e1023a6123e0005bf7">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="at_scale_schematic.png" class="img-fluid figure-img" width="750"></p>
<figcaption class="figure-caption">The two at-scale interfaces to <code>sceptre</code>. Users can analyze their data in the R console using the standard pipeline functions or on the command line using the Nextflow pipeline. In both cases users first import their data into an <code>ondisc</code>-backed <code>sceptre_object</code>, which is a kind of <code>sceptre_object</code> that stores the response and gRNA expression matrices on disk (as opposed to in memory).</figcaption></figure>
</div>
</div>
</div>
<p><code>ondisc</code> is an R package for large-scale computing on single-cell data, with a special focus on single-cell CRISPR screen analysis. <code>ondisc</code> implements several novel algorithms and data structures for efficiently manipulating and querying large single-cell expression matrices. <code>ondisc</code> is implemented as a lightweight layer of abstraction on top of the popular <a href="https://www.hdfgroup.org/solutions/hdf5/">HDF5</a> database system. Expression matrices within <code>ondisc</code> are stored on-disk in the custom <code>.odm</code> file format, which is an HDF5 file with special structure. <code>ondisc</code> enables fast, out-of-core access to the rows of a feature-by-cell expression matrix. (Thus, loading the expression vector of a given feature into memory is fast.) This chapter keeps discussion about <code>ondisc</code> to a minimum; interested readers can learn more about <code>ondisc</code> by reading XXX.</p>
<p>We use the high-MOI CRISPRi data that ship with <code>sceptre</code> as a running example in this chapter. We note that the example high-MOI CRISPRi data are fairly small; typically, we would leverage the at-scale interface of <code>sceptre</code> when the data are large (e.g., &gt;10 gigabytes). We describe each step of the Nextflow-based pipeline below, with one section per step. The title of each section indicates whether the step is to be carried out on a local machine (“local machine”) or on a cluster (“cluster”). By “local machine,” we mean a laptop or a desktop, and by “cluster,” we mean a computing cluster or a cloud.</p>
<p>We begin by loading the <code>sceptre</code> package.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-3_a25a39ba437a6f3e4abd73f7b85b2f24">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Katsevich-Lab/sceptre">sceptre</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sec-import_data_into_odm_backed_sceptre_object" class="level2" data-number="7.1"><h2 data-number="7.1" class="anchored" data-anchor-id="sec-import_data_into_odm_backed_sceptre_object">
<span class="header-section-number">7.1</span> Import data into an <code>ondisc</code>-backed <code>sceptre</code> object (local machine)</h2>
<p>The first step is to import the data into an <code>ondisc</code>-backed <code>sceptre</code> object. To this end, we call the function <code><a href="https://rdrr.io/pkg/sceptre/man/import_data_from_cellranger.html">import_data_from_cellranger()</a></code>, setting the argument <code>use_ondisc</code> to <code>TRUE</code>. Additionally, we set <code>directory_to_write</code> to a string indicating the directory in which to write the backing <code>.odm</code> files, which are the files that will contain the gene and gRNA expression matrices. In the example below we set <code>directory_to_write</code> to <code>"~/sceptre_data"</code>, although any valid file path will do. Finally, we set <code>moi</code> to the MOI of the dataset, <code>grna_target_data_frame</code> to the gRNA target data frame, and <code>directories</code> to the set of directories outputted by one or more calls to <code>cellranger count</code>. (The latter three arguments were described in depth in <a href="sceptre.html#sec-whole_game_import_data"><span>Section&nbsp;1</span></a> of <em>The whole game</em>.)</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-4_33efccf42767943c0307866274e24585">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># set the arguments</span></span>
<span><span class="va">directories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"sceptre"</span><span class="op">)</span>,</span>
<span>  <span class="st">"/highmoi_example/gem_group_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">grna_target_data_frame_highmoi</span><span class="op">)</span></span>
<span><span class="va">directory_to_write</span> <span class="op">&lt;-</span> <span class="st">"~/sceptre_data"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-6_00fbf270acc9ea78b10db0e423d77655">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># call the data import function</span></span>
<span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/import_data_from_cellranger.html">import_data_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories <span class="op">=</span> <span class="va">directories</span>,</span>
<span>  moi <span class="op">=</span> <span class="st">"high"</span>,</span>
<span>  grna_target_data_frame <span class="op">=</span> <span class="va">grna_target_data_frame_highmoi</span>,</span>
<span>  use_ondisc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">directory_to_write</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code><a href="https://rdrr.io/pkg/sceptre/man/import_data_from_cellranger.html">import_data_from_cellranger()</a></code> (with <code>use_ondisc</code> set to <code>TRUE</code>) converts the inputted <code>.mtx</code> files into <code>.odm</code> files, with one <code>.odm</code> file generated per modality. The <code>.odm</code> files contain the same information as the <code>.mtx</code> files, but the data are stored in a much more efficient format for single-cell CRISPR screen analysis. Additionally, <code><a href="https://rdrr.io/pkg/sceptre/man/import_data_from_cellranger.html">import_data_from_cellranger()</a></code> computes the cell-specific covariates (e.g., <code>response_n_umis</code>, <code>response_p_mito</code>, <code>batch</code>, etc.). Importantly, <code><a href="https://rdrr.io/pkg/sceptre/man/import_data_from_cellranger.html">import_data_from_cellranger()</a></code> uses only a few gigabytes of memory, even when the input data are tens of gigabytes in size.</p>
<p>Calling <code><a href="https://rdrr.io/r/base/list.files.html">list.files()</a></code> on <code>directory_to_write</code> shows that two files have been created in this directory: <code>gene.odm</code>, the file storing the gene expression matrix, and <code>grna.odm</code>, the file storing the gRNA expression matrix.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-8_ca596cb9be1396f4e6341470c04d1cf9">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">directory_to_write</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/timbarry/sceptre_data/gene.odm"
[2] "/Users/timbarry/sceptre_data/grna.odm"</code></pre>
</div>
</div>
</section><section id="explore-the-data-local-machine" class="level2" data-number="7.2"><h2 data-number="7.2" class="anchored" data-anchor-id="explore-the-data-local-machine">
<span class="header-section-number">7.2</span> Explore the data (local machine)</h2>
<p>The next step — which is optional but generally helpful — is to explore the data within the R console. There are three relevant functions in this context: <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_cell_covariates()</a></code>, <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_response_matrix()</a></code>, and <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_grna_matrix()</a></code>. <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_cell_covariates()</a></code> returns the cell-specific covariates that were computed as part of the data import step.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-9_5ebc6a7f0e0c586b70a58b6924b186e6">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell_covariates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_cell_covariates</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cell_covariates</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   response_n_umis response_n_nonzero response_p_mito grna_n_umis
             &lt;int&gt;              &lt;int&gt;           &lt;num&gt;       &lt;int&gt;
1:             830                171      0.04096386          24
2:             737                160      0.05563094           8
3:            1172                199      0.03412969           6
4:             787                168      0.04574333           8
5:             653                128      0.04594181          14
6:             879                177      0.02957907          25
   grna_n_nonzero   batch
            &lt;int&gt;  &lt;fctr&gt;
1:              3 batch_1
2:              1 batch_1
3:              1 batch_1
4:              1 batch_1
5:              1 batch_1
6:              1 batch_1</code></pre>
</div>
</div>
<p>Next, <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_response_matrix()</a></code> and <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_grna_matrix()</a></code> return the response and gRNA matrix, respectively. We call <code><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_response_matrix()</a></code> below.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-10_29c40a46f4b4a98440cc28bb9be3eae2">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">response_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_response_matrix</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>response_matrix</code> is an object of class <code>odm</code>, which is an integer-valued expression matrix backed by an <code>.odm</code> file. Evaluating <code>response_matrix</code> in the console prints the number of features and cells contained within the matrix, as well as the file path to the backing <code>.odm</code> file. The “Backing file” field is machine-specific.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-11_f5e2fe589f45d39f34de86b06654e5d2">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">response_matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class odm with the following attributes:
    • 526 features
    • 45919 cells
    • Backing file: /Users/timbarry/sceptre_data/gene.odm</code></pre>
</div>
</div>
<p><code>odm</code> objects support many of the key operators that standard R matrices support. For example, one can obtain the feature IDs of an <code>odm</code> object by calling <code><a href="https://rdrr.io/r/base/colnames.html">rownames()</a></code>.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-12_bcd9e701ac6f32bb92946915f64920f5">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">response_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">response_matrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">response_ids</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ENSG00000069275" "ENSG00000117222" "ENSG00000117266" "ENSG00000117280"
[5] "ENSG00000133059" "ENSG00000133065"</code></pre>
</div>
</div>
<p>Additionally, one can use the bracket operator (i.e, <code>[, ]</code>) to load a given row of the expression matrix into memory. For example, <code>response_matrix["ENSG00000069275",]</code> loads the expression vector corresponding to the gene “ENSG00000069275” into memory.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-13_01cc0cf8e82933a4e5de67a1b937d432">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_vector</span> <span class="op">&lt;-</span> <span class="va">response_matrix</span><span class="op">[</span><span class="st">"ENSG00000069275"</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expression_vector</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  4  6 10  9  4 13</code></pre>
</div>
</div>
<p><code>odm</code> objects take up very little space, as the expression data are stored on disk rather than in memory. For example, <code>response_matrix</code> takes up only 40 kilobytes (or 0.04 megabytes) of memory.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-14_40c240d87d7db3042f66a5a0a7baf603">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">response_matrix</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span><span class="op">(</span>units <span class="op">=</span> <span class="st">"Kb"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "38.6 Kb"</code></pre>
</div>
</div>
<p>Users can learn more about <code>odm</code> objects by referring to Appendix XXX.</p>
</section><section id="set-the-analysis-parameters-local-machine" class="level2" data-number="7.3"><h2 data-number="7.3" class="anchored" data-anchor-id="set-the-analysis-parameters-local-machine">
<span class="header-section-number">7.3</span> Set the analysis parameters (local machine)</h2>
<p>The third step is to set the analysis parameters. We proceed more or less exactly as described in <a href="sceptre.html#sec-whole_game_set_analysis_parameters"><span>Section&nbsp;2</span></a> of <em>The whole game</em>. We seek to conduct a <em>trans</em> analysis of the high-MOI CRISPRi data. Hence, we construct the positive control and discovery pairs using the functions <code><a href="https://rdrr.io/pkg/sceptre/man/construct_positive_control_pairs.html">construct_positive_control_pairs()</a></code> and <code><a href="https://rdrr.io/pkg/sceptre/man/construct_trans_pairs.html">construct_trans_pairs()</a></code>, respectively.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-15_42e63a608d671bf47800545ea49ab1ac">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># positive control pairs</span></span>
<span><span class="va">positive_control_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/construct_positive_control_pairs.html">construct_positive_control_pairs</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># discovery pairs</span></span>
<span><span class="va">discovery_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/construct_trans_pairs.html">construct_trans_pairs</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs</span>,</span>
<span>  pairs_to_exclude <span class="op">=</span> <span class="st">"pairs_containing_pc_targets"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we set the analysis parameters via a call to <code><a href="https://rdrr.io/pkg/sceptre/man/set_analysis_parameters.html">set_analysis_parameters()</a></code>. We set <code>resampling_mechanism</code> to <code>"permutations"</code> (as opposed to the default of <code>"crt"</code> in high-MOI), as <code>"permutations"</code> currently is the only option available for <code>resampling_mechanism</code> for an <code>ondisc</code>-backed <code>sceptre_object</code>. Additionally, we set <code>side</code> to <code>"both"</code> to conduct a two-tailed test, as we do not have a strong prior as to whether the perturbations will lead to an increase or decrease in the expression of the genes in <em>trans</em>.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-16_a4ac85e401bb68b2978f6986a910243d">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/set_analysis_parameters.html">set_analysis_parameters</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  discovery_pairs <span class="op">=</span> <span class="va">discovery_pairs</span>,</span>
<span>  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs</span>,</span>
<span>  side <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  resampling_mechanism <span class="op">=</span> <span class="st">"permutations"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We note that this step is optional; we instead could specify the analysis parameters on the command line as arguments to the Nextflow pipeline.</p>
</section><section id="write-the-ondisc-backed-sceptre_object-to-disk-local-machine" class="level2" data-number="7.4"><h2 data-number="7.4" class="anchored" data-anchor-id="write-the-ondisc-backed-sceptre_object-to-disk-local-machine">
<span class="header-section-number">7.4</span> Write the <code>ondisc</code>-backed <code>sceptre_object</code> to disk (local machine)</h2>
<p>The next step is to write the <code>ondisc</code>-backed <code>sceptre_object</code> to disk. To this end, we call the function <code><a href="https://rdrr.io/pkg/sceptre/man/read_ondisc_backed_sceptre_object.html">write_ondisc_backed_sceptre_object()</a></code>, passing <code>sceptre_object</code> and <code>directory_to_write</code> as arguments. The <code>sceptre_object</code> is saved within <code>directory_to_write</code> as <code>sceptre_object.rds</code>. <code>directory_to_write</code> can be any directory.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-17_6d088a23c1b57948ff9713d54a121845">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/read_ondisc_backed_sceptre_object.html">write_ondisc_backed_sceptre_object</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">directory_to_write</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>directory_to_write</code> now contains three files: <code>gene.odm</code>, <code>grna.odm</code>, and <code>sceptre_object.rds</code>.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-18_624d5505003d26f742d37464daf1516a">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">directory_to_write</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/timbarry/sceptre_data/gene.odm"          
[2] "/Users/timbarry/sceptre_data/grna.odm"          
[3] "/Users/timbarry/sceptre_data/sceptre_object.rds"</code></pre>
</div>
</div>
</section><section id="call-the-trial-nextflow-pipeline-local-machine" class="level2" data-number="7.5"><h2 data-number="7.5" class="anchored" data-anchor-id="call-the-trial-nextflow-pipeline-local-machine">
<span class="header-section-number">7.5</span> Call the trial Nextflow pipeline (local machine)</h2>
<p>The next step is to invoke the Nextflow pipeline on the local machine to analyze a small subset of the data. This small-scale analysis — which we call a “trial” analysis — enables us to verify that the pipeline is in working order. A Nextflow launch script for the example high-MOI data is as follows.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>launch_script.sh</strong></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Limit NF driver to 4 GB memory</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NXF_OPTS</span><span class="op">=</span><span class="st">"-Xms500M -Xmx4G"</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co"># REQUIRED INPUT ARGUMENTS</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="va">data_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_data/"</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sceptre object</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="va">sceptre_object_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"sceptre_object.rds"</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co"># response ODM</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="va">response_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"gene.odm"</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co"># grna ODM</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="va">grna_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"grna.odm"</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co">###################</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co"># OUTPUT DIRECTORY:</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="co">##################</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="va">output_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_outputs"</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Invoke pipeline</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run timothy-barry/sceptre-pipeline <span class="at">-r</span> main <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a> <span class="at">--sceptre_object_fp</span> <span class="va">$sceptre_object_fp</span> <span class="dt">\</span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a> <span class="at">--response_odm_fp</span> <span class="va">$response_odm_fp</span> <span class="dt">\</span></span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_odm_fp</span> <span class="va">$grna_odm_fp</span> <span class="dt">\</span></span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a> <span class="at">--output_directory</span> <span class="va">$output_directory</span> <span class="dt">\</span></span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_assignment_method</span> <span class="st">"mixture"</span> <span class="dt">\</span></span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a> <span class="at">--pair_pod_size</span> <span class="st">"30"</span> <span class="dt">\</span></span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_pod_size</span> <span class="st">"10"</span> <span class="dt">\</span></span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a> <span class="at">--trial</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will describe the structure of this launch script in detail shortly. For the time being, note that the <code>--trial</code> flag on the last line tells Nextflow to run the trial pipeline as opposed to the full-scale pipeline. Users can copy and paste the above script into a file called <code>launch_script.sh</code>. Alternately, users can access <code>launch_script.sh</code> by git cloning the <code>sceptre-pipeline</code> directory and then changing directories into the <code>examples</code> subdirectory, as follows.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>git clone git@github.com:timothy-barry/sceptre-pipeline.git
cd sceptre-pipeline/examples</code></pre>
</div>
<p>We run the pipeline by calling <code>bash</code> on <code>launch_script.sh</code>:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>bash launch_script.sh
# alternately: ./launch_script.sh</code></pre>
</div>
<p>If the pipeline runs correctly, the terminal output looks something like the following.</p>
<div class="cell" data-layout-align="center" data-hash="at-scale_cache/html/unnamed-chunk-19_6882a1a6cb17c3fe8709f0d198009f59">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="nf_output.png" class="img-fluid figure-img" width="750"></p>
<figcaption class="figure-caption">Nextflow output for the trial pipeline on a local machine.</figcaption></figure>
</div>
</div>
</div>
<p>The results of the pipeline are stored in <code>~/sceptre_outputs</code>. Calling <code>list.files</code> on <code>~/sceptre_outputs</code> shows that the directory contains all of the standard <code>sceptre</code> outputs, including the analysis summary text file (<code>analysis_summary.txt</code>), the plot files (e.g., <code>plot_assign_grnas.png</code>, etc.), and the result files (e.g., <code>results_run_discovery_analysis.rds</code>, etc.).</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-20_883090cf42a1aefe772bc02bb59a730e">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"~/sceptre_outputs"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "analysis_summary.txt"               "plot_assign_grnas.png"             
 [3] "plot_covariates.png"                "plot_grna_count_distributions.png" 
 [5] "plot_run_calibration_check.png"     "plot_run_discovery_analysis.png"   
 [7] "plot_run_power_check.png"           "plot_run_qc.png"                   
 [9] "results_run_calibration_check.rds"  "results_run_discovery_analysis.rds"
[11] "results_run_power_check.rds"       </code></pre>
</div>
</div>
<p>An inspection of <code>analysis_summary.txt</code> indicates that 97 negative control pairs and 97 discovery pairs were analyzed. The trial Nextflow pipeline randomly samples 100 pairs from the discovery set, 100 pairs from the negative control set, and 100 pairs from the positive control set and subjects these pairs to analysis. (Some of the pairs may be filtered out by pairwise QC. Moreover, if any of the sets contains fewer than 100 pairs, then that entire set of pairs is analyzed.) If the trial pipeline runs without issue, one can proceed to the next step.</p>
</section><section id="sec-push_data_to_cluster" class="level2" data-number="7.6"><h2 data-number="7.6" class="anchored" data-anchor-id="sec-push_data_to_cluster">
<span class="header-section-number">7.6</span> Push the data to the cluster</h2>
<p>The next step is to push the data to the cluster. The following block of code creates a directory <code>~/sceptre_data</code> on the cluster. It then copies the contents of <code>~/sceptre_data</code> (on the local machine) into <code>~/sceptre_data</code> (on the cluster). The strings “username” and “cluster_address” should be replaced by one’s username and cluster address, respectively.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>ssh username@cluster_address "mkdir -p ~/sceptre_data"
scp -r ~/sceptre_data username@cluster_address:~/</code></pre>
</div>
<p>We optionally can load and explore the data on the cluster. First, we ensure that the required packages are installed on the cluster.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-21_b4687798bd7906344eebfeb690b00062">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R, cluster</strong></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"katsevich-lab/sceptre"</span><span class="op">)</span> <span class="co"># sceptre</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"timothy-barry/ondisc"</span><span class="op">)</span> <span class="co"># ondisc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The function <code><a href="https://rdrr.io/pkg/sceptre/man/read_ondisc_backed_sceptre_object.html">read_ondisc_backed_sceptre_object()</a></code> loads an <code>ondisc</code>-backed <code>sceptre_object</code>, taking <code>sceptre_object_fp</code> (i.e., the file path to <code>sceptre_object.rds</code>), <code>response_odm_file_fp</code> (i.e., the file path to the response <code>.odm</code> file), and <code>grna_odm_file_fp</code> (i.e., the file path to the gRNA <code>.odm</code> file) as arguments. We briefly explore the <code>sceptre_object</code> to verify that looks the same on the cluster as on the local machine.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-22_3763d8077ff3311ca2d581e6d15a29ec">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R, cluster</strong></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Katsevich-Lab/sceptre">sceptre</a></span><span class="op">)</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">"~/sceptre_data/"</span></span>
<span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/read_ondisc_backed_sceptre_object.html">read_ondisc_backed_sceptre_object</a></span><span class="op">(</span> </span>
<span>  sceptre_object_fp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"sceptre_object.rds"</span><span class="op">)</span>,</span>
<span>  response_odm_file_fp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"gene.odm"</span><span class="op">)</span>,</span>
<span>  grna_odm_file_fp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"grna.odm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">response_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/get_response_matrix.html">get_response_matrix</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span>
<span><span class="va">expression_vector</span> <span class="op">&lt;-</span> <span class="va">response_matrix</span><span class="op">[</span><span class="st">"ENSG00000069275"</span>,<span class="op">]</span></span>
<span><span class="co"># etc...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We saved the example data to the home directory for convenience. On most clusters there is a dedicated “data” directory. Best practices dictate that users store their data in the dedicated data directory, especially when the data are large.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The first six steps of the Nextflow-based pipeline (i.e., <a href="#sec-import_data_into_odm_backed_sceptre_object"><span>Section&nbsp;7.1</span></a> – <a href="#sec-push_data_to_cluster"><span>Section&nbsp;7.6</span></a>) can be carried out entirely on the cluster, bypassing the local machine. If the output directories from <code>cellranger count</code> are stored on the cluster, the user can import the data into an <code>ondisc</code>-backed <code>sceptre_object</code>, set the analysis parameters, and write the updated <code>sceptre_object</code> to disk on the cluster as follows.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-23_6d6076b09e690effa5b2498246668a2d">
<details><summary>Show code</summary><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># on cluster</span></span>
<span><span class="co"># import data</span></span>
<span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/import_data_from_cellranger.html">import_data_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories <span class="op">=</span> <span class="va">directories</span>,</span>
<span>  moi <span class="op">=</span> <span class="st">"high"</span>,</span>
<span>  grna_target_data_frame <span class="op">=</span> <span class="va">grna_target_data_frame_highmoi</span>,</span>
<span>  use_ondisc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">directory_to_write</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># set analysis parameters (define discovery_pairs, positive_control_pairs)</span></span>
<span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/set_analysis_parameters.html">set_analysis_parameters</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  discovery_pairs <span class="op">=</span> <span class="va">discovery_pairs</span>,</span>
<span>  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs</span>,</span>
<span>  side <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  resampling_mechanism <span class="op">=</span> <span class="st">"permutations"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># write updated sceptre_object to disk</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/sceptre/man/read_ondisc_backed_sceptre_object.html">write_ondisc_backed_sceptre_object</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">directory_to_write</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The above script should require only a few (e.g., 3 – 4) gigabytes of memory, even if the input data are tens of gigabytes in size. This approach, though streamlined, is less interactive than the local-machine-based approach outlined in sections <a href="#sec-import_data_into_odm_backed_sceptre_object"><span>Section&nbsp;7.1</span></a> – <a href="#sec-push_data_to_cluster"><span>Section&nbsp;7.6</span></a>.</p>
</div>
</div>
</section><section id="call-the-trial-pipeline-cluster" class="level2" data-number="7.7"><h2 data-number="7.7" class="anchored" data-anchor-id="call-the-trial-pipeline-cluster">
<span class="header-section-number">7.7</span> Call the trial pipeline (cluster)</h2>
<p>The next step is to call the trial pipeline on the cluster. First, we download the <code>launch_script.sh</code> submission script onto the cluster using the same command as before.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal, cluster</strong></pre>
</div>
<pre class="terminal"><code>git clone git@github.com:timothy-barry/sceptre-pipeline.git
cd sceptre-pipeline/examples</code></pre>
</div>
<p>We launch the submission script by submitting the script to the scheduler. For example, on a sun grid engine (SGE) cluster, one would use the <code>qsub</code> command.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal, cluster</strong></pre>
</div>
<pre class="terminal"><code>qsub launch_script.sh</code></pre>
</div>
<p>On SLURM cluster, one would use the <code>sbatch</code> command.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal, cluster</strong></pre>
</div>
<pre class="terminal"><code>sbatch launch_script.sh</code></pre>
</div>
<p>Note that we are able to launch the pipeline using essentially the same command on our local machine as on the cluster. This is one of the advantages of Nextflow: it provides a layer of abstraction between the logic of the pipeline and the execution environment. Thus, we can develop (or debug, etc.) a pipeline on our local machine and then run it at scale on the cluster without modification.</p>
<p>The pipeline outputs are written to the directory <code>~/sceptre_outputs</code>. Again, all of the standard <code>sceptre</code> outputs — e.g., <code>analysis_summary.txt</code>, <code>results_run_discovery_analysis.rds</code>, etc. — are contained within this directory. In fact, the <code>~/sceptre_outputs</code> directory on the cluster should match exactly the <code>~/sceptre_outputs</code> directory on the local machine. We can view <code>analysis_summary.txt</code>, for example, by calling <code>cat</code> on this file:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal, cluster</strong></pre>
</div>
<pre class="terminal"><code>cat analysis_summary.txt</code></pre>
</div>
</section><section id="call-the-full-pipeline-cluster" class="level2" data-number="7.8"><h2 data-number="7.8" class="anchored" data-anchor-id="call-the-full-pipeline-cluster">
<span class="header-section-number">7.8</span> Call the full pipeline (cluster)</h2>
<p>The final step is to run the full pipeline on the cluster. To this end, we modify <code>launch_script.sh</code>, removing the <code>--trial</code> flag from the last line of the script. We can do this on the cluster by calling <code>nano launch_script.sh</code>, deleting the <code>--trial</code> flag, and saving the updated script. <code>launch_script.sh</code> now should look as follows.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-24_b03c47ccaf023478ecc47f072a076dda">
<details><summary>Show code</summary><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Limit NF driver to 4 GB memory</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NXF_OPTS</span><span class="op">=</span><span class="st">"-Xms500M -Xmx4G"</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co"># REQUIRED INPUT ARGUMENTS</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="va">data_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_data/"</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sceptre object</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a><span class="va">sceptre_object_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"sceptre_object.rds"</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="co"># response ODM</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a><span class="va">response_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"gene.odm"</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="co"># grna ODM</span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a><span class="va">grna_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"grna.odm"</span></span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a><span class="co">###################</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a><span class="co"># OUTPUT DIRECTORY:</span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="co">##################</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="va">output_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_outputs"</span></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Invoke pipeline</span></span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run timothy-barry/sceptre-pipeline <span class="at">-r</span> main <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a> <span class="at">--sceptre_object_fp</span> <span class="va">$sceptre_object_fp</span> <span class="dt">\</span></span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a> <span class="at">--response_odm_fp</span> <span class="va">$response_odm_fp</span> <span class="dt">\</span></span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_odm_fp</span> <span class="va">$grna_odm_fp</span> <span class="dt">\</span></span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a> <span class="at">--output_directory</span> <span class="va">$output_directory</span> <span class="dt">\</span></span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_assignment_method</span> <span class="st">"mixture"</span> <span class="dt">\</span></span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a> <span class="at">--pair_pod_size</span> <span class="st">"30"</span> <span class="dt">\</span></span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_pod_size</span> <span class="st">"10"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Finally, we can launch the full pipeline by calling <code>qsub</code> (or <code>sbatch</code>, depending on the scheduler) on <code>launch_script.sh</code>:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal, cluster</strong></pre>
</div>
<pre class="terminal"><code>qsub launch_script.sh
# or sbatch launch_script.sh</code></pre>
</div>
<p>The outputs are written to <code>~/sceptre_outputs</code>. (The previous outputs are overwritten.) We can download the output directory onto our local machine as follows.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal, local</strong></pre>
</div>
<pre class="terminal"><code>scp -r username@cluster_address:~/sceptre_outputs ~/</code></pre>
</div>
<p>Again, “username” and “cluster_address” should be replaced by one’s username and cluster address, respectively. An exploration of <code>~/sceptre_outputs</code> on the local machine shows that we have successfully carried out a <em>trans</em> analysis of the example high-MOI CRISPRi data.</p>
</section><section id="stepping-through-the-pipeline-incrementally" class="level2" data-number="7.9"><h2 data-number="7.9" class="anchored" data-anchor-id="stepping-through-the-pipeline-incrementally">
<span class="header-section-number">7.9</span> Stepping through the pipeline incrementally</h2>
<p>The <code>sceptre</code> Nextflow pipeline consists of six distinct steps, depicted below.</p>
<div class="cell" data-layout-align="center" data-hash="at-scale_cache/html/unnamed-chunk-25_e2f39ac3cf2ef5c5320085abacd907cd">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="reduced_pipeline.png" class="img-fluid figure-img" width="500"></p>
<figcaption class="figure-caption">The <code>sceptre</code> Nextflow pipeline.</figcaption></figure>
</div>
</div>
</div>
<p>The Nextflow pipeline is nearly the same as the pipeline introduced in <a href="sceptre.html#sec-whole_game_import_data"><span>Section&nbsp;1</span></a> — <a href="sceptre.html#sec-sceptre_write_outputs_to_directory"><span>Section&nbsp;8</span></a> of <em>The whole game.</em> The difference is that the “Import data” step is not part of this pipeline (as we import the data into an <code>ondisc</code>-backed <code>sceptre</code> object prior to calling the Nextflow pipeline), and likewise the “Write outputs to directory” step is not part of this pipeline (as the Nextflow pipeline automatically writes results to the output directory at the end of each step).</p>
<p>We can advance through the Nextflow pipeline one step at a time by specifying the <code>pipeline_stop</code> argument. <code>pipeline_stop</code> takes six possible values: <code>"set_analysis_parameters"</code>, <code>"assign_grnas"</code>, <code>"run_qc"</code>, <code>"run_calibration_check"</code>, <code>"run_power_check"</code>, and <code>"run_discovery_analysis"</code>. When <code>pipeline_stop</code> is specified, the Nextflow pipeline runs through all the steps of the pipeline up to and including <code>pipeline_stop</code>. For example, if we set <code>pipeline_stop</code> to <code>"run_calibration_check"</code>, the Nextflow pipeline carries out the steps “Set analysis parameters,” “Assign gRNAs,” “Run quality control,” and “Run calibration check” and then stops. The default value of <code>pipeline_stop</code> is <code>"run_discovery_analysis"</code>, which is why the Nextflow pipeline by default runs through to completion.</p>
<p>We provide an example of using the <code>pipeline_stop</code> argument below. This example can be carried out either on the local machine or on the cluster. First, let us remove the <code>~/sceptre_outputs</code> directory.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>rm -r ~/sceptre_outputs</code></pre>
</div>
<p>We modify <code>launch_script.sh</code>, adding the line <code>--pipeline_stop "assign_grnas"</code> to the script. This line tells Nextflow to stop at the “Assign gRNAs” step.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-26_21855f874fc5a278d37af0c520257170">
<details><summary>Show code</summary><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Limit NF driver to 4 GB memory</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NXF_OPTS</span><span class="op">=</span><span class="st">"-Xms500M -Xmx4G"</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="co"># REQUIRED INPUT ARGUMENTS</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="va">data_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_data/"</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sceptre object</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="va">sceptre_object_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"sceptre_object.rds"</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="co"># response ODM</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="va">response_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"gene.odm"</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a><span class="co"># grna ODM</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a><span class="va">grna_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"grna.odm"</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a><span class="co">###################</span></span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a><span class="co"># OUTPUT DIRECTORY:</span></span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a><span class="co">##################</span></span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a><span class="va">output_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_outputs"</span></span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Invoke pipeline</span></span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run timothy-barry/sceptre-pipeline <span class="at">-r</span> main <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a> <span class="at">--sceptre_object_fp</span> <span class="va">$sceptre_object_fp</span> <span class="dt">\</span></span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a> <span class="at">--response_odm_fp</span> <span class="va">$response_odm_fp</span> <span class="dt">\</span></span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_odm_fp</span> <span class="va">$grna_odm_fp</span> <span class="dt">\</span></span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a> <span class="at">--output_directory</span> <span class="va">$output_directory</span> <span class="dt">\</span></span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_assignment_method</span> <span class="st">"mixture"</span> <span class="dt">\</span></span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a> <span class="at">--pair_pod_size</span> <span class="st">"30"</span> <span class="dt">\</span></span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_pod_size</span> <span class="st">"10"</span> <span class="dt">\</span></span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a> <span class="at">--pipeline_stop</span> <span class="st">"assign_grnas"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We then launch the pipeline (via <code>bash launch_script.sh</code>, <code>qsub launch_script.sh</code>, or <code>sbatch launch_script.sh</code>, etc.) and examine the outputs directory. We see (via an inspection of <code>analysis_summary.txt</code>, for example) that “Set analysis parameters” and “Assign gRNAs” have completed, but that the other pipeline steps have not yet run.</p>
<p>Suppose now that want to run the next step in the pipeline, namely “Run quality control.” To this end, we update <code>launch_script.sh</code>, changing the line <code>--pipeline_stop "assign_grnas"</code> to <code>--pipeline_stop "run_qc"</code>. Additionally, we add the line <code>-resume</code> to the launch script. The <code>-resume</code> flag in effect tells Nextflow to skip over all steps preceeding “Run QC” (including “Assign gRNAs”), as these steps already have been run.</p>
<div class="cell" data-hash="at-scale_cache/html/unnamed-chunk-27_cf15323dc573722716e4b31fc95f50b9">
<details><summary>Show code</summary><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Limit NF driver to 4 GB memory</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NXF_OPTS</span><span class="op">=</span><span class="st">"-Xms500M -Xmx4G"</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># REQUIRED INPUT ARGUMENTS</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co">##########################</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="va">data_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_data/"</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sceptre object</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="va">sceptre_object_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"sceptre_object.rds"</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co"># response ODM</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="va">response_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"gene.odm"</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a><span class="co"># grna ODM</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="va">grna_odm_fp</span><span class="op">=</span><span class="va">$data_directory</span><span class="st">"grna.odm"</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a><span class="co">###################</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a><span class="co"># OUTPUT DIRECTORY:</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a><span class="co">##################</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a><span class="va">output_directory</span><span class="op">=</span><span class="va">$HOME</span><span class="st">"/sceptre_outputs"</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Invoke pipeline</span></span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a><span class="co">#################</span></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run timothy-barry/sceptre-pipeline <span class="at">-r</span> main <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a> <span class="at">--sceptre_object_fp</span> <span class="va">$sceptre_object_fp</span> <span class="dt">\</span></span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a> <span class="at">--response_odm_fp</span> <span class="va">$response_odm_fp</span> <span class="dt">\</span></span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_odm_fp</span> <span class="va">$grna_odm_fp</span> <span class="dt">\</span></span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a> <span class="at">--output_directory</span> <span class="va">$output_directory</span> <span class="dt">\</span></span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_assignment_method</span> <span class="st">"mixture"</span> <span class="dt">\</span></span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a> <span class="at">--pair_pod_size</span> <span class="st">"30"</span> <span class="dt">\</span></span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a> <span class="at">--grna_pod_size</span> <span class="st">"10"</span> <span class="dt">\</span></span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a> <span class="at">--pipeline_stop</span> <span class="st">"run_qc"</span> <span class="dt">\</span></span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a> <span class="at">-resume</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center" data-hash="at-scale_cache/html/unnamed-chunk-28_8c36d627cff1d4b5da589476dd3ebe17">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="nf_output_2.png" class="img-fluid figure-img" width="700"></p>
<figcaption class="figure-caption">Nextflow output when the <code>-resume</code> flag is used.</figcaption></figure>
</div>
</div>
</div>
<p>We launch the pipeline and display the terminal output above. The “Assign gRNAs” step has not been recomputed, as indicated by the “cached” string next to the <code>assign_grnas</code> process. In conclusion, <code>--pipeline_stop</code> and <code>-resume</code> enable one to step through the Nextflow pipeline incrementally, verifying that a given step of the pipeline has run correctly before proceeding to the next step. If a given step of the pipeline produces an unexpected or suboptimal result, one can update the arguments passed to that step of the pipeline to try to improve the result.</p>
</section><section id="sec-install_nextflow" class="level2" data-number="7.10"><h2 data-number="7.10" class="anchored" data-anchor-id="sec-install_nextflow">
<span class="header-section-number">7.10</span> Addendum: installing Nextflow</h2>
<p>Nextflow is a popular (and increasingly dominant) tool for orchestrating complex bioinformatic workflows. Users will need to install Nextflow on their local computer and cluster to use the <code>sceptre</code> Nextflow pipeline. The <a href="https://www.nextflow.io/docs/latest/index.html">Nextflow website</a> contains extensive documentation, including instructions on downloading and installing Nextflow. Briefly, you can install Nextflow as follows.</p>
<ol type="1">
<li>Install SDKMAN!, a system for installing and managing software.</li>
</ol>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>curl -s "https://get.sdkman.io" | bash</code></pre>
</div>
<ol start="2" type="1">
<li>Install Java.</li>
</ol>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>sdk install java 17.0.6-amzn</code></pre>
</div>
<p>If this fails, try the following instead.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>sdk install java 17.0.6-tem</code></pre>
</div>
<p>You can ensure that Java is installed via the following command.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>java -version </code></pre>
</div>
<p>You should see something like the output below. Ensure that the version of Java (as printed on the first line of the output) is greater than or equal to 17.0.0. (Note: the minimum version of Java required by Nextflow may have increased since the time of writing.)</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>openjdk version "17.0.6" 2023-01-17 LTS
OpenJDK Runtime Environment Corretto-17.0.6.10.1 (build 17.0.6+10-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.6.10.1 (build 17.0.6+10-LTS, mixed mode, sharing)</code></pre>
</div>
<ol start="3" type="1">
<li>Create a directory in which to store the Nextflow executable. We create a directory called <code>bin</code> in our home directory for this purpose, although the Nextflow executable can be stored in any directory.</li>
</ol>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>mkdir -p ~/bin</code></pre>
</div>
<ol start="4" type="1">
<li>Change directories to <code>~/bin</code> and intall the Nextflow binary.</li>
</ol>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>cd ~/bin
wget -qO- https://get.nextflow.io | bash
# alternately: curl -s https://get.nextflow.io | bash</code></pre>
</div>
<p>This creates a binary called <code>nextflow</code> in the current directory.</p>
<ol start="5" type="1">
<li>Make the <code>nextflow</code> binary an executable by running the following command.</li>
</ol>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>chmod +x nextflow</code></pre>
</div>
<p>(Note: The installed binary may already be executable; in this case <code>chmod</code> does nothing.)</p>
<ol start="6" type="1">
<li>Add <code>nextflow</code> to your linux path. One way to do this is as follows. First, open <code>.bash_profile</code> (via, for example, <code>nano ~/.bash_profile</code>). Next, add the following line to <code>.bash_profile</code>: <code>export PATH="~/bin/:$PATH"</code>. Finally, save <code>.bash_profile</code>, close the terminal, and open a new terminal. When you enter <code>nextflow -version</code> on the command line, you should see an output similar to the following.</li>
</ol>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>terminal</strong></pre>
</div>
<pre class="terminal"><code>N E X T F L O W
version 23.10.1 build 5891
created 12-01-2024 22:01 UTC (17:01 EDT)
cite doi:10.1038/nbt.3820
http://nextflow.io</code></pre>
</div>
<p>This indicates that Nextflow has been installed successfully.</p>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./run-power-check-and-discovery-analysis.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Run power check and discovery analysis</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./pipeline-details.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Details of the pipeline</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Hands-On Single-Cell CRISPR Screen Analysis</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>