<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Timothy Barry, Eugene Katsevich">
<title>Hands-On Single-Cell CRISPR Screen Analysis - Appendix A — The ondisc package</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./sceptredata.html" rel="next">
<link href="./methods.html" rel="prev">
<link href="./book_cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="custom.css">
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./ondisc.html">Appendices</a></li><li class="breadcrumb-item"><a href="./ondisc.html"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">The <code>ondisc</code> package</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./hex.jpg" alt="" class="sidebar-logo py-0 d-lg-inline d-none"></a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Hands-On Single-Cell CRISPR Screen Analysis</a> 
        <div class="sidebar-tools-main">
    <a href="https://katsevich-lab.github.io/sceptre/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-git"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sceptre.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The whole game</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Part I. Introduction to sceptre</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./import-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Import data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./set-analysis-parameters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Set analysis parameters</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assign-grnas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Assign gRNAs</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./run-qc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Run quality control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./run-calibration-check.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Run calibration check</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./run-power-check-and-discovery-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Run power check and discovery analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Part II. At-scale sceptre</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./at-scale.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to the Nextflow pipeline</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipeline-details.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">A deeper dive into the Nextflow pipeline</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipeline-args.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Nextflow installation, pipeline arguments</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Part III. Methods and algorithms</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Overview of methods</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ondisc.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">The <code>ondisc</code> package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sceptredata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">The <code>sceptredata</code> package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./glossary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Glossary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">D</span>&nbsp; <span class="chapter-title">Frequently asked questions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">E</span>&nbsp; <span class="chapter-title">References</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#initializing-an-odm-object-via-create_odm_from_cellranger" id="toc-initializing-an-odm-object-via-create_odm_from_cellranger" class="nav-link active" data-scroll-target="#initializing-an-odm-object-via-create_odm_from_cellranger"><span class="header-section-number">A.1</span> Initializing an <code>odm</code> object via <code>create_odm_from_cellranger()</code></a></li>
  <li><a href="#interacting-with-the-odm-object" id="toc-interacting-with-the-odm-object" class="nav-link" data-scroll-target="#interacting-with-the-odm-object"><span class="header-section-number">A.2</span> Interacting with the <code>odm</code> object</a></li>
  <li><a href="#supported-modalities" id="toc-supported-modalities" class="nav-link" data-scroll-target="#supported-modalities"><span class="header-section-number">A.3</span> Supported modalities</a></li>
  <li><a href="#the-cell-wise-covariate-data-frame" id="toc-the-cell-wise-covariate-data-frame" class="nav-link" data-scroll-target="#the-cell-wise-covariate-data-frame"><span class="header-section-number">A.4</span> The cell-wise covariate data frame</a></li>
  <li><a href="#reading-an-.odm-file-into-r" id="toc-reading-an-.odm-file-into-r" class="nav-link" data-scroll-target="#reading-an-.odm-file-into-r"><span class="header-section-number">A.5</span> Reading an <code>.odm</code> file into R</a></li>
  <li><a href="#initializing-an-odm-object-via-create_odm_from_r_matrix" id="toc-initializing-an-odm-object-via-create_odm_from_r_matrix" class="nav-link" data-scroll-target="#initializing-an-odm-object-via-create_odm_from_r_matrix"><span class="header-section-number">A.6</span> Initializing an <code>odm</code> object via <code>create_odm_from_r_matrix()</code></a></li>
  <li><a href="#notes-on-compression" id="toc-notes-on-compression" class="nav-link" data-scroll-target="#notes-on-compression"><span class="header-section-number">A.7</span> Notes on compression</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./ondisc.html">Appendices</a></li><li class="breadcrumb-item"><a href="./ondisc.html"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">The <code>ondisc</code> package</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-ondisc" class="quarto-section-identifier">Appendix A — The <code>ondisc</code> package</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p><a href="https://timothy-barry.github.io/ondisc/"><code>ondisc</code></a> is a companion R package to <code>sceptre</code> that facilitates analysis of large-scale single-cell data out-of-core on a laptop or distributed across tens to hundreds processors on a cluster or cloud. In both of these settings, <code>ondisc</code> requires only a few gigabytes of memory, even if the input data are tens of gigabytes in size. <code>ondisc</code> mainly is oriented toward single-cell CRISPR screen analysis, but <code>ondisc</code> also can be used for single-cell differential expression and single-cell co-expression analyses. <code>ondisc</code> is powered by several new, efficient algorithms for manipulating and querying large, sparse expression matrices. Although <code>ondisc</code> and <code>sceptre</code> work best in conjunction, <code>ondisc</code> can be used independently of <code>sceptre</code> (and conversely, <code>sceptre</code> can be used independently of <code>ondisc</code>).</p>
<p>Users can install <code>ondisc</code> using the code below. <code>ondisc</code> depends on the Bioconductor package <code>Rhdf5lib</code>, which should be installed from source before installing <code>ondisc</code>. Users also should install <code>sceptredata</code>, which contains the example data used in this vignette.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># install.packages("BiocManager"); install.packages("devtools")</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"Rhdf5lib"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span> <span class="co"># Rhdf5lib</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"timothy-barry/ondisc"</span><span class="op">)</span> <span class="co"># ondisc</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"katsevich-lab/sceptredata"</span><span class="op">)</span> <span class="co"># sceptredata</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>See the <a href="./faq.html">frequently asked questions page</a> for tips on installing <code>ondisc</code> such that it runs as fast as possible. We can load <code>ondisc</code> and <code>sceptredata</code> by calling <code><a href="https://rdrr.io/r/base/library.html">library()</a></code>.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://timothy-barry.github.io/ondisc/">ondisc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://timothy-barry.github.io/sceptre-book/">sceptredata</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The interface to <code>ondisc</code> is simple and minimal. The package contains only one class: <code>odm</code> (short for “<code>ondisc</code> matrix”). An <code>odm</code> object represents a single-cell expression matrix stored <em>on disk</em> (as opposed to <em>in memory</em>). <code>odm</code> objects can be used to store expression matrices that are too large to fit in memory. Users can create an <code>odm</code> object via one of two functions: <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> or <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>. The former takes the output of one or more calls to Cell Ranger count as input, while the latter takes an R matrix (stored in standard format or sparse format) as input. Users can interface with an <code>odm</code> object using several functions, including the bracket (<code>[,]</code>) operator, which loads a specified subset of the expression matrix into memory.</p>
<section id="initializing-an-odm-object-via-create_odm_from_cellranger" class="level2" data-number="A.1"><h2 data-number="A.1" class="anchored" data-anchor-id="initializing-an-odm-object-via-create_odm_from_cellranger">
<span class="header-section-number">A.1</span> Initializing an <code>odm</code> object via <code>create_odm_from_cellranger()</code>
</h2>
<p><code>ondisc</code> provides two functions for initializing an <code>odm</code> object: <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> and <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>. The former is considerably more scalable and memory-efficient than the latter; thus, we recommend that users employ <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> when possible. We illustrate use of <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> on an example single-cell CRISPR screen dataset stored in the <code>sceptredata</code> package. The example data contain two modalities, namely a gene modality and a CRISPR gRNA modality. There are 526 genes, 95 gRNAs, and 45,919 cells in the data. Users can read more about the example data by evaluating <code><a href="https://bioconductor.org/packages/release/bioc/vignettes/sceptredata/inst/doc/sceptredata.html">vignette("sceptredata")</a></code> or <code><a href="https://rdrr.io/pkg/sceptredata/man/highmoi_example_data.html">?highmoi_example_data</a></code> in the console. <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> takes several arguments: <code>directories_to_load</code>, <code>directory_to_write</code>, <code>write_cellwise_covariates</code>, <code>chunk_size</code>, <code>compression_level</code>, and <code>grna_target_data_frame</code>. Only the first two of these arguments are required; the rest are set to reasonable defaults. We describe the <code>directories_to_load</code> and <code>directory_to_write</code> arguments below.</p>
<p><code>directories_to_load</code> is a character vector specifying the locations of one or more directories outputted by Cell Ranger count. Below, we set <code>directories_to_load</code> to the (machine-specific) location of the example data on disk.</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">directories_to_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"sceptredata"</span><span class="op">)</span>, </span>
<span>  <span class="st">"/highmoi_example/gem_group_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">directories_to_load</span> <span class="co"># file paths to the example data on your computer</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/sceptredata/extdata/highmoi_example/gem_group_1"
[2] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/sceptredata/extdata/highmoi_example/gem_group_2"</code></pre>
</div>
</div>
<p><code>directories_to_load</code> contains the file paths to two directories, which correspond to cells sequenced across two batches. The data are stored in <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-mex-matrices">feature barcode format</a>; each directory contains the files <code>barcodes.tsv.gz</code>, <code>features.tsv.gz</code>, and <code>matrix.mtx.gz</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">directories_to_load</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "barcodes.tsv.gz" "features.tsv"    "matrix.mtx"     </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">directories_to_load</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"  </code></pre>
</div>
</div>
<p>Next, <code>directory_to_write</code> is a file path to the directory in which to write the backing <code>.odm</code> file, which is the file that will store the expression data on disk. <code>.odm</code> files contain the same information as <code>.mtx</code> files but stored in a more efficient format for CRISPR screen analysis, differential expression analysis, and gene co-expression analysis. <code>.odm</code> files simply are HDF5 files with special structure. We set <code>directory_to_write</code> to <code>temp_dir</code> (i.e., the temporary directory) in this example. The remaining arguments are optional, and most users will not need to specify them; see <code>?create_odm_from_cellranger()</code> for more information. Below, we call <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> on the example data, saving the output of the function to the variable <code>out_list</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories_to_load <span class="op">=</span> <span class="va">directories_to_load</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">temp_dir</span> </span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>out_list</code> contains three entries: <code>gene</code>, <code>grna</code>, and <code>cellwise_covariates</code>. <code>gene</code> and <code>grna</code> are the <code>odm</code> objects corresponding to the gene and gRNA modalities, respectively. Meanwhile, <code>cellwise_covariates</code> is a data frame that contains the cell-wise covariates. (More on the cell-wise covariates later.) An inspection of <code>temp_dir</code> reveals that the files <code>gene.odm</code> and <code>grna.odm</code> have been written to this directory.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">temp_dir</span>, pattern <span class="op">=</span> <span class="st">"*.odm"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gene.odm" "grna.odm"</code></pre>
</div>
</div>
</section><section id="interacting-with-the-odm-object" class="level2" data-number="A.2"><h2 data-number="A.2" class="anchored" data-anchor-id="interacting-with-the-odm-object">
<span class="header-section-number">A.2</span> Interacting with the <code>odm</code> object</h2>
<p>We extract the <code>odm</code> object corresponding to the gene modality as follows.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="va">out_list</span><span class="op">[[</span><span class="st">"gene"</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Evaluating an <code>odm</code> object in the console prints information about the matrix, including the number of features and cells contained within the matrix, as well as the file path to the (machine-specific) backing <code>.odm</code> file.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_odm</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class odm with the following attributes:
    • 526 features
    • 45919 cells
    • Backing file: /var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpmiQpuB/gene.odm</code></pre>
</div>
</div>
<p><code>odm</code> objects support several key matrix operations, including <code><a href="https://rdrr.io/r/base/nrow.html">ncol()</a></code>, <code><a href="https://rdrr.io/r/base/nrow.html">nrow()</a></code>, <code><a href="https://rdrr.io/r/base/colnames.html">rownames()</a></code>, and <code>[,]</code>. <code><a href="https://rdrr.io/r/base/nrow.html">ncol()</a></code> and <code><a href="https://rdrr.io/r/base/nrow.html">nrow()</a></code> return the number of rows (i.e., features) and columns (i.e., cells) contained within the matrix, respectively.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span>
<span><span class="va">n_features</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 526</code></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span>
<span><span class="va">n_cells</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 45919</code></pre>
</div>
</div>
<p>Next, <code><a href="https://rdrr.io/r/base/colnames.html">rownames()</a></code> returns the feature IDs.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">feature_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">feature_ids</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ENSG00000069275" "ENSG00000117222" "ENSG00000117266" "ENSG00000117280"
[5] "ENSG00000133059" "ENSG00000133065"</code></pre>
</div>
</div>
<p>Finally, the bracket operator (<code>[,]</code>) loads a specified row of the expression matrix into memory. One can index into the rows by integer index or feature ID, as follows.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_vector</span> <span class="op">&lt;-</span> <span class="va">gene_odm</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expression_vector</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 1 0 1 1 0</code></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_vector</span> <span class="op">&lt;-</span> <span class="va">gene_odm</span><span class="op">[</span><span class="st">"ENSG00000117222"</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expression_vector</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 1 0 1 1 0</code></pre>
</div>
</div>
<p>Indexing into an <code>odm</code> object by column is not supported. Finally, <code>odm</code> objects take up very little space, as the data are stored on-disk rather than in-memory. For example, <code>gene_odm</code> takes up only 40 kilobytes of memory.</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span><span class="op">(</span>units <span class="op">=</span> <span class="st">"Kb"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "38.7 Kb"</code></pre>
</div>
</div>
</section><section id="supported-modalities" class="level2" data-number="A.3"><h2 data-number="A.3" class="anchored" data-anchor-id="supported-modalities">
<span class="header-section-number">A.3</span> Supported modalities</h2>
<p><code>ondisc</code> supports the following Cell Ranger modalities: <code>Gene Expression</code>, <code>CRISPR Guide Capture</code> (i.e., gRNA expression), and <code>Antibody Capture</code> (i.e., protein expression). (The modality of a given feature is listed within the third column of the unzipped <code>features.tsv</code> file; see the <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-mex-matrices">Cell Ranger documentation</a> for more information.) The table below maps the modality name used by Cell Ranger to that used by <code>ondisc</code>.</p>
<table class="table">
<thead><tr class="header">
<th>Cell Ranger modality name</th>
<th>
<code>ondisc</code> modality name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>Gene Expression</code></td>
<td><code>gene</code></td>
</tr>
<tr class="even">
<td><code>CRISPR Guide Capture</code></td>
<td><code>grna</code></td>
</tr>
<tr class="odd">
<td><code>Antibody Capture</code></td>
<td><code>protein</code></td>
</tr>
</tbody>
</table>
<p>We provide an example of using <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> to import a dataset containing three modalities: gene expression, gRNA expression, and protein expression. We use a synthetic dataset for this purpose (so as to reduce the amount of data stored within the <code>sceptredata</code> package). To this end we call the function <code><a href="https://timothy-barry.github.io/ondisc/reference/write_example_cellranger_dataset.html">write_example_cellranger_dataset()</a></code>, which creates a synthetic single-cell dataset, writing the dataset to disk in Cell Ranger feature barcode format. (See <code>?write_example_cellranger_dataset()</code> for more information about this function.) We create a synthetic single-cell dataset consisting of 500 genes, 50 gRNAs, 20 proteins, and 10,000 cells. Furthermore, we specify that the cells are sequenced across three batches. We write the synthetic dataset to the directory <code>temp_dir</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://timothy-barry.github.io/ondisc/reference/write_example_cellranger_dataset.html">write_example_cellranger_dataset</a></span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">50</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>  n_cells <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  n_batch <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  modalities <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"grna"</span>, <span class="st">"protein"</span><span class="op">)</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">temp_dir</span> ,</span>
<span>  p_set_col_zero <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The synthetic data are contained in the directories <code>batch_1</code>, <code>batch_2</code>, and <code>batch_3</code> within <code>temp_dir</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">directories_to_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span></span>
<span>  <span class="va">temp_dir</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"batch_"</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">directories_to_load</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpmiQpuB/batch_1"
[2] "/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpmiQpuB/batch_2"
[3] "/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpmiQpuB/batch_3"</code></pre>
</div>
</div>
<p>Each of these directories contains the files <code>matrix.mtx.gz</code>, <code>features.tsv.gz</code>, and <code>barcodes.tsv.gz</code>. For example, the contents of the <code>batch_1</code> are as follows.</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">directories_to_load</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"  </code></pre>
</div>
</div>
<p>We call <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> to import these data, saving the output of the function in the variable <code>out_list</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">out_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories_to_load <span class="op">=</span> <span class="va">directories_to_load</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">temp_dir</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>out_list</code> contains the cell-wise covariate data frame alongside <code>odm</code> objects corresponding to the gene, gRNA, and protein modalities.</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">out_list</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gene"                "grna"                "protein"            
[4] "cellwise_covariates"</code></pre>
</div>
</div>
<p>Moreover, the files <code>gene.odm</code>, <code>grna.odm</code>, and <code>protein.odm</code> have been written to disk. (The previous <code>gene.odm</code> and <code>grna.odm</code> files are overwritten.)</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">temp_dir</span>, pattern <span class="op">=</span> <span class="st">"*.odm"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gene.odm"    "grna.odm"    "protein.odm"</code></pre>
</div>
</div>
</section><section id="the-cell-wise-covariate-data-frame" class="level2" data-number="A.4"><h2 data-number="A.4" class="anchored" data-anchor-id="the-cell-wise-covariate-data-frame">
<span class="header-section-number">A.4</span> The cell-wise covariate data frame</h2>
<p>As part of importing the data, <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> computes the cell-wise covariates. We print the first few rows of the cell-wise covariate data frame corresponding to the synthetic data below.</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cellwise_covariates</span> <span class="op">&lt;-</span> <span class="va">out_list</span><span class="op">[[</span><span class="st">"cellwise_covariates"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cellwise_covariates</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene_n_umis gene_n_nonzero gene_p_mito grna_n_umis grna_n_nonzero
         &lt;int&gt;          &lt;int&gt;       &lt;num&gt;       &lt;int&gt;          &lt;int&gt;
1:        1030            196   0.4330097         131             22
2:        1034            187   0.4197292         126             24
3:        1142            203   0.4168126         126             20
4:        1177            217   0.4188615         119             21
5:        1083            207   0.4524469         118             24
6:        1095            193   0.4000000          89             17
   grna_feature_w_max_expression grna_frac_umis_max_feature protein_n_umis
                          &lt;char&gt;                      &lt;num&gt;          &lt;int&gt;
1:                       grna_33                 0.07633588             64
2:                        grna_6                 0.07936508             46
3:                       grna_31                 0.07936508             51
4:                       grna_22                 0.08403361             44
5:                       grna_20                 0.08474576             31
6:                       grna_11                 0.11235955             56
   protein_n_nonzero   batch
               &lt;int&gt;  &lt;fctr&gt;
1:                 9 batch_1
2:                 9 batch_1
3:                 8 batch_1
4:                 8 batch_1
5:                 7 batch_1
6:                 9 batch_1</code></pre>
</div>
</div>
<p>The modality to which a given covariate corresponds (“gene”, “grna”, or “protein”) is prepended to the name of the covariate. We describe each covariate below.</p>
<ul>
<li><p><code>gene_n_umis</code>: the number of gene UMIs sequenced in a given cell.</p></li>
<li><p><code>gene_n_nonzero</code>: the number of genes that exhibit nonzero expression in a given cell.</p></li>
<li><p><code>gene_p_mito</code>: the fraction of gene transcripts that map to mitochondrial genes in a given cell. (Mitochondrial genes are identified as genes whose name starts with <code>"MT-"</code> or <code>"mt-"</code>.)</p></li>
<li><p><code>grna_n_umis</code>: similar to <code>gene_n_umis</code> but for the gRNA modality.</p></li>
<li><p><code>grna_n_nonzero</code>: similar to <code>gene_n_nonzero</code> but for the gRNA modality.</p></li>
<li><p><code>grna_feature_w_max_expression</code>: the ID of the gRNA that exhibits the maximum UMI count in a given cell.</p></li>
<li><p><code>grna_frac_umis_max_feature</code>: the fraction of UMIs that the maximally expressed gRNA in a given cell constitutes.</p></li>
<li><p><code>protein_n_umis</code>: similar to <code>gene_n_umis</code> but for the protein modality.</p></li>
<li><p><code>protein_n_nonzero</code>: similar to <code>gene_n_nonzero</code> but for the protein modality.</p></li>
<li><p><code>batch</code>: the batch in which a given cell was sequenced. Cells loaded from different directories are assumed to belong to different batches.</p></li>
</ul>
<p><code>sceptre</code> uses the covariates <code>grna_feature_w_max_expression</code> and <code>grna_frac_umis_max_feature</code> to assign gRNAs to cells.</p>
</section><section id="reading-an-.odm-file-into-r" class="level2" data-number="A.5"><h2 data-number="A.5" class="anchored" data-anchor-id="reading-an-.odm-file-into-r">
<span class="header-section-number">A.5</span> Reading an <code>.odm</code> file into R</h2>
<p>Users can read an <code>.odm</code> file into R by calling the function <code><a href="https://timothy-barry.github.io/ondisc/reference/initialize_odm_from_backing_file.html">initialize_odm_from_backing_file()</a></code>. Below, we delete all variables from the global namespace. Then, we call <code><a href="https://timothy-barry.github.io/ondisc/reference/initialize_odm_from_backing_file.html">initialize_odm_from_backing_file()</a></code> on the file <code>gene.odm</code> stored within <code>temp_dir</code>, which loads the gene expression matrix that we created in the previous step.</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># delete all variables</span></span>
<span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://timothy-barry.github.io/ondisc/reference/initialize_odm_from_backing_file.html">initialize_odm_from_backing_file</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"/gene.odm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gene_odm</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class odm with the following attributes:
    • 500 features
    • 10000 cells
    • Backing file: /var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpmiQpuB/gene.odm</code></pre>
</div>
</div>
<p><code>.odm</code> files are portable. Thus, a user can create an <code>.odm</code> file on one computer, move the <code>.odm</code> file to another computer, and then open the <code>.odm</code> file on the second computer.</p>
</section><section id="initializing-an-odm-object-via-create_odm_from_r_matrix" class="level2" data-number="A.6"><h2 data-number="A.6" class="anchored" data-anchor-id="initializing-an-odm-object-via-create_odm_from_r_matrix">
<span class="header-section-number">A.6</span> Initializing an <code>odm</code> object via <code>create_odm_from_r_matrix()</code>
</h2>
<p>We recommend that users create an <code>odm</code> object via <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code>, as this function is highly scalable and typically requires only a couple gigabytes of memory. However, users also can convert an R matrix into an <code>odm</code> object via the function <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>. <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> takes two main arguments: <code>mat</code> and <code>file_to_write</code>. <code>mat</code> is a standard R matrix (of type <code>"matrix"</code>) or a sparse R matrix (of type <code>"dgCMatrix"</code>, <code>"dgRMatrix"</code>, or <code>"dgTMatrix"</code>). <code>mat</code> should contain row names giving the ID of each feature. Next, <code>file_to_write</code> is a fully-qualified file path specifying the location in which to write the backing <code>.odm</code> file. We provide an example of calling <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> on a gene-by-cell expression matrix contained in the <code>sceptredata</code> package.</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">lowmoi_example_data</span><span class="op">)</span></span>
<span><span class="va">gene_mat</span> <span class="op">&lt;-</span> <span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">response_matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>gene_mat</code> is a gene expression matrix containing 299 genes and 20,729 cells. (Users can evaluate <code><a href="https://rdrr.io/pkg/sceptredata/man/lowmoi_example_data.html">?lowmoi_example_data</a></code> to see more information about this matrix.) We pass this matrix to <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>, setting <code>file_to_write</code> to <code>paste0(temp_dir, "/gene.odm")</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_to_write</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"/gene.odm"</span><span class="op">)</span></span>
<span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">gene_mat</span>,</span>
<span>  file_to_write <span class="op">=</span> <span class="va">file_to_write</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>gene_odm</code> is a standard <code>odm</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_odm</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class odm with the following attributes:
    • 299 features
    • 20729 cells
    • Backing file: /var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpmiQpuB/gene.odm</code></pre>
</div>
</div>
<p>Moreover, the file <code>gene.odm</code> has been written to <code>temp_dir</code>. (The previous <code>gene.odm</code> file is overwritten.)</p>
</section><section id="notes-on-compression" class="level2" data-number="A.7"><h2 data-number="A.7" class="anchored" data-anchor-id="notes-on-compression">
<span class="header-section-number">A.7</span> Notes on compression</h2>
<p><code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> and <code><a href="https://timothy-barry.github.io/ondisc/reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> take optional arguments <code>chunk_size</code> and <code>compression_level</code> (which are set to reasonable defaults). <code>chunk_size</code> and <code>compression_level</code> control the extent to which the backing <code>.odm</code> file is compressed. <code>chunk_size</code> should be a positive integer, and <code>compression_level</code> should be an integer in the range of 0 to 9. Increasing the value of these arguments <em>increases</em> the level of compression, thereby leading to a smaller file size for the backing <code>.odm</code> file (but possibly longer read and write times).</p>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./methods.html" class="pagination-link  aria-label=" of="" methods="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Overview of methods</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./sceptredata.html" class="pagination-link" aria-label="<span class='chapter-number'>B</span>&nbsp; <span class='chapter-title'>The `sceptredata` package</span>">
        <span class="nav-page-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">The <code>sceptredata</code> package</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Hands-On Single-Cell CRISPR Screen Analysis</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>


</body></html>